src/Ads/AccountService.php

```diff
@@ -160,14 +161,28 @@ public function setup_account(): array {
 						$this->check_billing_status( $account );
 						break;
 
-					case 'link_merchant':
-						$this->link_merchant_account();
-						break;
-
 					case 'conversion_action':
 						$this->create_conversion_action();
 						break;
 
+					case 'link_merchant':
+						$merchant_id = $this->options->get_merchant_id();
+						// As MC and Ads account can be connected interchangeably, this check is necessary.
+						if ( ! empty( $ads_id ) && ! empty( $merchant_id ) ) {
+							$this->link_merchant_account();
+						} else {
+							/**
+							 * Mark the step as pending, save the state and
+							 * continue the foreach loop with `continue 2`.
+							 *
+							 * Marking the step as pending will make sure to run this step next time.
+							 */
+							$state[ $name ]['status'] = AdsAccountState::STEP_PENDING;
+							$this->state->update( $state );
+							continue 2;
+						}
+						break;
```

Comment:
This pattern is still confusing, which is made obvious by `test_setup_account_step_link_merchant_no_ads_id`. When we don't have a Merchant ID we want to hit the else condition and set the state as pending, but when we are setting up an ads account without an ads ID, that's a clear error scenario. Since we already have that being thrown within `link_merchant_account`, I'd suggest to refactor this code to make it flow better (so we need less comments to explain what's happening):

```suggestion
					case 'link_merchant':
						// Continue to next step if the MC account is not connected yet.
						if ( ! $this->options->get_merchant_id() ) {
							// Save step as pending and continue the foreach loop with `continue 2`.
							$state[ $name ]['status'] = AdsAccountState::STEP_PENDING;
							$this->state->update( $state );
							continue 2;
						}

						$this->link_merchant_account();
						break;
```

I've also removed the empty check since `get_merchant_id` will always return an `int`. As a followup to this we'll need to revert the test `test_setup_account_step_link_merchant_no_ads_id`